File: sourcecode/server/controllers/resumeController.js
const AWS = require('aws-sdk');
const { v4: uuidv4 } = require('uuid');

AWS.config.update({
  accessKeyId: process.env.AWS_ACCESS_KEY_ID,
  secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
  region: process.env.AWS_REGION
});

const s3 = new AWS.S3();

const uploadResume = async (req, res) => {
    try {
      console.log('Upload request received');
      console.log('Request body:', req.body);
      console.log('Request file:', req.file ? {
        originalname: req.file.originalname,
        mimetype: req.file.mimetype,
        size: req.file.size
      } : 'No file');

      const file = req.file;
      // Check if job description was provided
      const jobDescription = req.body.jobDescription || '';

      console.log('Job description from request body:', jobDescription ?
        (jobDescription.length > 100 ? jobDescription.substring(0, 100) + '...' : jobDescription) :
        'None provided');

      // Generate a unique ID for this upload
      const uploadId = uuidv4();
      const key = `textract-output/${uploadId}-${file.originalname}`;

      // Prepare S3 upload parameters
      const params = {
        Bucket: process.env.AWS_S3_BUCKET,
        Key: key,
        Body: file.buffer,
        ContentType: file.mimetype
      };

      // Add job description as metadata if provided
      // Note: S3 metadata has limitations - ASCII characters only and size limits
      if (jobDescription) {
        // For simplicity and reliability, we'll store a flag in metadata
        // and the actual job description as a separate object if it's too long
        if (jobDescription.length > 1000) {
          // If job description is too long, store it as a separate object
          console.log('Job description is too long for metadata, storing as separate object');

          // Store job description in S3
          const jobDescKey = `job-descriptions/${uploadId}.txt`;
          await s3.upload({
            Bucket: process.env.AWS_S3_BUCKET,
            Key: jobDescKey,
            Body: jobDescription,
            ContentType: 'text/plain'
          }).promise();

          // Add a reference to the job description in metadata
          params.Metadata = {
            'has-job-description': 'true',
            'job-description-key': jobDescKey
          };
        } else {
          // If job description is short enough, store it directly in metadata
          // Ensure it only contains ASCII characters and is URL-encoded
          // S3 metadata has strict requirements
          const asciiJobDesc = encodeURIComponent(jobDescription.replace(/[^\x00-\x7F]/g, ''));
          console.log('ASCII and URL-encoded job description:', asciiJobDesc.substring(0, 100) + '...');

          params.Metadata = {
            'job-description': asciiJobDesc
          };
        }
      }

      if (jobDescription) {
        console.log('Job description provided, length:', jobDescription.length);
        console.log('Adding job description as metadata');
        console.log('First 100 chars of job description:', jobDescription.substring(0, 100));

        // Log the params object to verify metadata is set correctly
        console.log('S3 upload params:', JSON.stringify({
          Bucket: params.Bucket,
          Key: params.Key,
          ContentType: params.ContentType,
          Metadata: params.Metadata
        }));
      } else {
        console.log('No job description provided');
      }

      console.log('S3 upload params:', {
        Bucket: params.Bucket,
        Key: params.Key,
        ContentType: params.ContentType,
        Metadata: params.Metadata
      });

      const data = await s3.upload(params).promise();
      console.log('S3 upload response:', JSON.stringify(data));

      // Verify the metadata was set correctly by getting the object metadata
      try {
        const headResponse = await s3.headObject({
          Bucket: params.Bucket,
          Key: params.Key
        }).promise();

        console.log('S3 object metadata after upload:', headResponse.Metadata);
      } catch (headError) {
        console.error('Error getting object metadata after upload:', headError);
      }

      res.json({ message: 'File uploaded successfully', data });
    } catch (error) {
      console.error('Upload error:', error);
      res.status(500).json({ error: 'Error uploading file' });
    }
  };


  const getRewrittenResume = async (req, res) => {
    try {
      const { key } = req.query; // frontend will send the key

      console.log('Fetching rewritten resume with key:', key);

      if (!key) {
        return res.status(400).json({ error: 'Missing key parameter' });
      }

      // First try to find the exact file with the key
      try {
        const params = {
          Bucket: process.env.AWS_S3_BUCKET,
          Key: `rewritten-resumes/${key}`, // Try the exact key first
        };

        console.log('Trying exact key S3 params:', params);
        const data = await s3.getObject(params).promise();
        console.log('S3 data received with exact key, content length:', data.Body.length);

        const jsonContent = JSON.parse(data.Body.toString('utf-8'));
        console.log('JSON parsed successfully');

        res.setHeader('Content-Type', 'application/json');
        return res.send(jsonContent);
      } catch (exactKeyError) {
        console.log('Exact key not found, trying to find file with key in the name...');

        // If exact key fails, try to find a file that contains the key in its name
        try {
          // List all files in the rewritten-resumes folder
          const listParams = {
            Bucket: process.env.AWS_S3_BUCKET,
            Prefix: 'rewritten-resumes/'
          };

          const listData = await s3.listObjectsV2(listParams).promise();
          console.log(`Found ${listData.Contents.length} objects in the bucket`);

          // Filter out the folder itself and only return actual files
          const files = listData.Contents.filter(item => !item.Key.endsWith('/'));
          console.log(`Found ${files.length} files after filtering`);

          // Try to find a file that contains the key in its name
          const matchingFile = files.find(file => file.Key.includes(key));

          if (matchingFile) {
            console.log('Found matching file:', matchingFile.Key);

            const getParams = {
              Bucket: process.env.AWS_S3_BUCKET,
              Key: matchingFile.Key
            };

            const fileData = await s3.getObject(getParams).promise();
            console.log('File content received, size:', fileData.Body.length);

            const jsonContent = JSON.parse(fileData.Body.toString('utf-8'));
            console.log('JSON parsed successfully');

            res.setHeader('Content-Type', 'application/json');
            return res.send(jsonContent);
          } else {
            // If no matching file is found, return the latest file
            console.log('No matching file found, returning the latest file...');

            // Sort by LastModified date (newest first)
            files.sort((a, b) => new Date(b.LastModified) - new Date(a.LastModified));

            if (files.length > 0) {
              // Get the most recent file
              const latestFile = files[0];
              console.log('Latest file:', latestFile.Key, 'Last modified:', latestFile.LastModified);

              const getParams = {
                Bucket: process.env.AWS_S3_BUCKET,
                Key: latestFile.Key
              };

              const fileData = await s3.getObject(getParams).promise();
              console.log('Latest file content received, size:', fileData.Body.length);

              const jsonContent = JSON.parse(fileData.Body.toString('utf-8'));
              console.log('JSON parsed successfully');

              res.setHeader('Content-Type', 'application/json');
              return res.send(jsonContent);
            } else {
              return res.status(404).json({
                error: 'No resumes found in the bucket',
                details: `No files found in rewritten-resumes/ in bucket ${process.env.AWS_S3_BUCKET}`
              });
            }
          }
        } catch (listError) {
          console.error('Error listing or processing files:', listError);
          throw listError;
        }
      }
    } catch (error) {
      console.error('Fetch rewritten resume error:', error);
      res.status(500).json({
        error: 'Error fetching resume',
        message: error.message
      });
    }
  };


module.exports = { uploadResume, getRewrittenResume };

File: sourcecode/server/controllers/resume_controller.py
import os
import json
import logging
import uuid
import boto3
from flask import jsonify
from werkzeug.utils import secure_filename

# Configure logging
logging.basicConfig(level=logging.INFO,
                    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

# Configure AWS S3
s3 = boto3.client(
    's3',
    aws_access_key_id=os.environ.get('AWS_ACCESS_KEY_ID'),
    aws_secret_access_key=os.environ.get('AWS_SECRET_ACCESS_KEY'),
    region_name=os.environ.get('AWS_REGION', 'us-east-1')
)

def upload_resume(request):
    """
    Upload a resume file to S3
    """
    try:
        if 'file' not in request.files:
            return jsonify({'error': 'No file part'}), 400

        file = request.files['file']
        if file.filename == '':
            return jsonify({'error': 'No selected file'}), 400

        # Generate a unique filename
        unique_id = str(uuid.uuid4())
        original_filename = secure_filename(file.filename)
        key = f"textract-output/{unique_id}-{original_filename}"

        # Upload to S3
        s3.upload_fileobj(
            file,
            os.environ.get('AWS_S3_BUCKET'),
            key,
            ExtraArgs={
                'ContentType': file.content_type
            }
        )

        return jsonify({
            'message': 'File uploaded successfully',
            'data': {
                'Bucket': os.environ.get('AWS_S3_BUCKET'),
                'Key': key
            }
        })
    except Exception as error:
        logger.error(f'Upload error: {error}')
        return jsonify({'error': 'Error uploading file'}), 500

def get_rewritten_resume(request):
    """
    Get a rewritten resume from S3 by key
    """
    try:
        key = request.args.get('key')
        logger.info(f'Fetching rewritten resume with key: {key}')

        if not key:
            return jsonify({'error': 'Missing key parameter'}), 400

        # First try to find the exact file with the key
        try:
            params = {
                'Bucket': os.environ.get('AWS_S3_BUCKET'),
                'Key': f"rewritten-resumes/{key}"  # Try the exact key first
            }

            logger.info(f'Trying exact key S3 params: {params}')
            data = s3.get_object(**params)
            # Read the content once and store it
            file_content = data["Body"].read()
            logger.info(f'S3 data received with exact key, content length: {len(file_content)}')

            json_content = json.loads(file_content.decode('utf-8'))
            logger.info('JSON parsed successfully')

            return jsonify(json_content)
        except Exception as exact_key_error:
            logger.info('Exact key not found, trying to find file with key in the name...')

            # If exact key fails, try to find a file that contains the key in its name
            try:
                # List all files in the rewritten-resumes folder
                list_params = {
                    'Bucket': os.environ.get('AWS_S3_BUCKET'),
                    'Prefix': 'rewritten-resumes/'
                }

                list_data = s3.list_objects_v2(**list_params)
                logger.info(f'Found {len(list_data.get("Contents", []))} objects in the bucket')

                # Filter out the folder itself and only return actual files
                files = [item for item in list_data.get('Contents', []) if not item['Key'].endswith('/')]
                logger.info(f'Found {len(files)} files after filtering')

                # Try to find a file that contains the key in its name
                matching_file = next((file for file in files if key in file['Key']), None)

                if matching_file:
                    logger.info(f'Found matching file: {matching_file["Key"]}')

                    get_params = {
                        'Bucket': os.environ.get('AWS_S3_BUCKET'),
                        'Key': matching_file['Key']
                    }

                    file_data = s3.get_object(**get_params)
                    # Read the content once and store it
                    file_content = file_data["Body"].read()
                    logger.info(f'File content received, size: {len(file_content)}')

                    json_content = json.loads(file_content.decode('utf-8'))
                    logger.info('JSON parsed successfully')

                    return jsonify(json_content)
                else:
                    # If no matching file is found, return the latest file
                    logger.info('No matching file found, returning the latest file...')

                    # Sort by LastModified date (newest first)
                    files.sort(key=lambda x: x['LastModified'], reverse=True)

                    if files:
                        # Get the most recent file
                        latest_file = files[0]
                        logger.info(f'Latest file: {latest_file["Key"]}, Last modified: {latest_file["LastModified"]}')

                        get_params = {
                            'Bucket': os.environ.get('AWS_S3_BUCKET'),
                            'Key': latest_file['Key']
                        }

                        file_data = s3.get_object(**get_params)
                        # Read the content once and store it
                        file_content = file_data["Body"].read()
                        logger.info(f'Latest file content received, size: {len(file_content)}')

                        json_content = json.loads(file_content.decode('utf-8'))
                        logger.info('JSON parsed successfully')

                        return jsonify(json_content)
                    else:
                        return jsonify({
                            'error': 'No resumes found in the bucket',
                            'details': f'No files found in rewritten-resumes/ in bucket {os.environ.get("AWS_S3_BUCKET")}'
                        }), 404
            except Exception as list_error:
                logger.error(f'Error listing or processing files: {list_error}')
                raise list_error
    except Exception as error:
        logger.error(f'Fetch rewritten resume error: {error}')
        return jsonify({
            'error': 'Error fetching resume',
            'message': str(error)
        }), 500
